Coding system versions: part 1

rebkwok opened the pull request:
The beginning of an implementaion of coding system versioning - more info in [this issue](https://github.com/opensafely-core/opencodelists/issues/1379) and [doc](https://docs.google.com/document/d/1rstaBR4u6MpLfH4urfDA27mZyVowMuvrkUE7kOZFVuM/edit).

This PR implements the first stage of [this proposal](https://gist.github.com/rebkwok/65bf170765910ed5846ef3c4311a6023), whereby each new version of a coding system will live in its own (read-only) sqlite database. It sets up the new CodelistSystemVersion model, and preparations for coding systems to live in separate databases, but it doesn't change anything about how opencodelists works yet.

More specifically changes are:
1) A new `coding_systems.versioning` app with one model, `CodingSystemVersion`, to hold information about a single import of a version of one coding system.   The intention is that each  CodingSystemVersion instance will correspond to a separate database holding the coding system data for that version.

2) A function `update_coding_system_database_connections`, called from the `versioning` app's `AppConfig.ready` method, which iterates through all the `CodingSystemVersion` instances and adds the relevant database config to django's database connections.  Note that we can't add these to the `DATABASES` in `settings.py` because we can't access the model instances until the apps are fully loaded.  (An alternative would be to use the sqlite files in the filesystem - we could do that in settings.py)

3) A management command that extracts the current coding system tables from the main DB, creates an initial `CodingSystemVersion` instance for each coding system (with an "unknown" version, since this will be assigned initially to all the current CodelistVersions that pre-date the coding system versioning), and loads the tables into separate databases.

inglesp reviewed (commented):
Changes look reasonable, but before merging I'd like to know what the plan is for linking an existing `CodelistVersion` against a `CodingSystemVersion`.

I'd also like to know whether you've thought about whether the tests should use a separate database for each coding system version.

A couple of things that would have made this a little easier to review:

* rebasing so that code introduced in one commit isn't changed (eg functions renamed or moved) in another one in the same PR
* adding tests for new behaviour in the same commit as the new behaviour

rebkwok commented:
> Changes look reasonable, but before merging I'd like to know what the plan is for linking an existing CodelistVersion against a CodingSystemVersion.
Working on that now - draft PR #1401 .  Currently I've made a CodelistVersion have a charfield with a `coding_system_version_hash` rather than a FK.  The coding_system,py module for each coding system version is now a class that takes the db alias. This minimises the code changes that we need in the rest of the codebase, because calls that use `codelist_version.coding_system` still have the same interface.

> I'd also like to know whether you've thought about whether the tests should use a separate database for each coding system version.
Also looking at that - that was my original intention, but pytest-django's support for multiple databases is experimental and I can't get the fixtures that use `loaddata` to respect the `databases` parameter - they just ignore it and load into the default database.  The tests in this PR do test the database switching, so my current thought is to have an `override_db` field on the `CodingSystemVersion` that would let a  `CodingSystemVersion` in a test use the default db.

rebkwok commented on coding_systems/versioning/models.py:
```diff
@@ -0,0 +1,39 @@
+
+class CodingSystemVersion(models.Model):
+    coding_system = models.CharField(max_length=10)
+    version = models.CharField(max_length=255, null=True, blank=True)
```
I think I had an idea initially that  this would be an official version identifier and they might not all have one, but I guess even if they didn't, we should label them with one anyway

inglesp replied:
> we should label them with one anyway

I agree.

rebkwok commented on coding_systems/versioning/management/commands/migrate_coding_system.py:
```diff
@@ -0,0 +1,124 @@
+
+        for coding_system in coding_systems:
+            # Check this coding system is migrate-able (i.e. it's a valid app)
+            if coding_system not in apps.app_configs:
```
yes, there's `null` and `opcs4` coding systems that aren't in `INSTALLED_APPS`

rebkwok commented on coding_systems/versioning/management/commands/migrate_coding_system.py:
```diff
@@ -0,0 +1,124 @@
+
+            self.load_data_to_new_database(coding_system, new_db_path)
+
+    def create_new_database(self, coding_system_version, force):
```
ah, no.  It did it an earlier incarnation

rebkwok replied:
Function name and docstring updated now

rebkwok commented on coding_systems/versioning/management/commands/migrate_coding_system.py:
```diff
@@ -0,0 +1,124 @@
+
+            new_db_path = self.create_new_database(coding_system_version, force)
+
+            self.load_data_to_new_database(coding_system, new_db_path)
```
I've put that in a data migration in the next PR

rebkwok commented on coding_systems/versioning/management/commands/migrate_coding_system.py:
```diff
@@ -0,0 +1,132 @@
+
+            new_db_path = self.create_new_database(coding_system_version, force)
+
+            self.load_data_to_new_database(coding_system, new_db_path, source_db)
```
I thought yes, but not just yet...

rebkwok commented on coding_systems/versioning/tests/test_models.py:
```diff
@@ -0,0 +1,66 @@
+    assert CodingSystemVersion.latest("snomedct") == coding_system_version
+
+    # an unknown coding system
+    assert CodingSystemVersion.latest("foo") is None
```
I wasn't sure - a `Codelist` has a `coding_system` property too, which I think should return the most recent version.  But it can legitimately be None, if the codelist is opcs4 one, so it either needs to check if it's None, or catch an expected exception.  Is the latter bettter?

rebkwok commented on coding_systems/versioning/models.py:
```diff
@@ -0,0 +1,60 @@
+
+    @property
+    def db_name(self):
+        return f"{self.coding_system}_{self.hash}"
```
Updated now so the db_name is just a SlugField set on the model, generated from the coding system, version and valid_from date, which should all always be present and unique.

rebkwok commented on coding_systems/versioning/tests/test_models.py:
```diff
@@ -0,0 +1,66 @@
+    with pytest.raises(ConnectionDoesNotExist):
+        Concept.objects.using(coding_system_version.db_name).exists()
+
+    update_coding_system_database_connections()
```
This is done now, with a fixture that removes the new database connections added by this function after the tests run.  It's in the top-level conftest.py because this method gets called whenever the versioning app is loaded, so it could potentially be modified in any test.

rebkwok commented:
@inglesp I think I have addressed all comments now, apart from the one here about raising an exception vs returning None.  Happy to update that if you think it's better that way, as I couldn't decide.
https://github.com/opensafely-core/opencodelists/pull/1397#discussion_r1001962240

Also, linking the CodelistVersion to the CodingSystemVersion - I did that in the next PR in a data migration that depends on the new CodingSystemVersions and their databases having been created.  I left it out of this one because I didn't want to add anything yet that made any change to the current functionality (also why the coding system data isn't deleted from the main database in this PR).  I could put the CodelistVersion linking into this one if you prefer - just adding a coding system version as a field on the CodelistVersion wouldn't change any functionality, and it probably does make sense to do it alongside migrating the coding system data.

rebkwok commented:
Note: I have now figured out (in the next PR) how to ensure the tests are also using the separate coding system databases

inglesp commented on coding_systems/versioning/models.py:
```diff
@@ -2,39 +2,38 @@
+    @cached_property
     def db_name(self):
-        return f"{self.coding_system}_{self.hash}"
+        return self.slug
```
Do we need both `db_name` and `slug`?

rebkwok replied:
Originally I just called the SlugField "db_name", but then I changed my mind and thought it was more intuitive to name it "slug" and alias it to "db_name".  There were places where I was retrieving the `CodingSystemVersion` and it didn't feel quite right to be retrieving it by db_name.  I'm not committed to it though

inglesp replied:
I've no strong opinion.
